app/models/user.rb

```diff
@@ -940,6 +940,189 @@ def remove_image(image)
+  # licenses, we are perfectly within our rights to retain users' content.
+  # But we wish to comply with Apple where it will not otherwise detract from
+  # other users' experience.
+  def disable_account_and_delete_private_objects
+    disable_account
+    delete_api_keys
+    delete_interests
+    delete_notifications
+    delete_queued_emails
+    delete_observations
+    delete_private_name_descriptions
+    delete_private_location_descriptions
+    delete_private_projects
+    delete_private_species_lists
+    delete_unattached_collection_numbers
+    delete_unattached_herbarium_records
+    delete_unattached_images
+    User.erase_user(id) if no_references_left?
+  end
+
+  # Disable and remove all public information from account but leave it there
+  # in case there are still comments, etc. on the site by this user.
+  def disable_account
+    self.password = ""
+    self.email = ""
+    self.blocked = true
+    self.location_id = nil
+    self.image_id = nil
+    self.notes = nil
+    self.mailing_address = nil
+    self.save
+  end
+
+  def delete_api_keys
+    api_keys.delete_all
+  end
+
+  def delete_interests
+    interests.delete_all
+  end
+
+  def delete_notifications
+    notifications.delete_all
+  end
+
+  def delete_queued_emails
+    QueuedEmail.where(user_id: id).delete_all
+    QueuedEmail.where(to_user_id: id).delete_all
+  end
+
+  def delete_observations
+    [Naming, Vote, RssLog].each do |model|
+      model.joins(:observation).where(observation: { user_id: id }).delete_all
+    end
+    # (all the rest of the observations' dependents should autodestruct)
+    observations.delete_all
+  end
+
+  # Delete user's descriptions that don't have any other authors or # editors.
+  def delete_private_name_descriptions
+    ids = (name_descriptions -
+             name_descriptions.joins(:name_description_authors).
+               where.not(name_description_authors: { user_id: id }) -
+             name_descriptions.joins(:name_description_editors).
+               where.not(name_description_editors: { user_id: id })).
+          map(&:id)
+    NameDescription.where(id: ids).delete_all
+  end
+
+  # Delete user's descriptions that don't have any other authors or # editors.
+  def delete_private_location_descriptions
+    ids = (location_descriptions -
+            location_descriptions.joins(:location_description_authors).
+              where.not(location_description_authors: { user_id: id }) -
+            location_descriptions.joins(:location_description_editors).
+              where.not(location_description_editors: { user_id: id })).
+          map(&:id)
+    LocationDescription.where(id: ids).delete_all
+  end
+
+  # Delete all the user's projects that don't have any other users on them.
+  def delete_private_projects
+    Project.joins(:admin_group, :user_group).
+      where(user: self,
+            admin_group: { name: "user #{id}" },
+            user_group: { name: "user #{id}" }).
```

Comment:
This is short and it works.
But perhaps is a bit hacky; it depends on the convention for naming a user's personal UserGroup.
See `UserGroup#self.create_user`.
IMO we should get quickly to something that works so that the iOS mobile app can be published.
We can refactor later.

Reply:
I didn't like this because there may be other user groups with just that one user in it.  In fact the test fixtures bolete_admins and bolete_users both contain only Dick!  I think I found a better way to do it which is more in line with all the other related methods.

Also, on a related note, I noticed that the test had been rewritten to basically use the exact same logic as the method it was testing (it looked like a copy and paste from the method), so of course if the logic is wrong in the code, then the test won't catch it.  I realize the motivation is to make the test more robust with respect to changing fixtures, but it also makes the test less useful, much harder to understand, and well, prone to being outright wrong without realizing it!  (Such as in this case.)  I rewrote it hard-coded the way you presumably hate, but compromised by documenting better what all the requirements are that are being tested and making sure they are all met before I even begin the actual testing itself.  At least this way when the test breaks because of someone monkeying around with the fixtures, it should be easier to understand how to modify the test to accommodate the changes.

(I personally think the better solution to this problem, instead of writing more complex and potentially buggy tests, is to leave the fixtures alone!  Instead, if you need some very particular setup for one or two tests, create it on the fly in those tests, and not in the application-wide fixtures.  That way the details of the new fixtures are clearly "documented" where they are needed (and you don't have to go searching ad nauseum through the fixtures to understand them), and they won't break any other tests.  Everyone wins.  Except maybe it's a bit slower.  I am a very strong fan of keeping tests as logic-free as possible.  Tests do not need to be DRY.  But that's just my two cents.)

Reply:
Thanks for the re-write. I won't have time to look at it.
I'm good with more hard-coding in the tests (and also testing the fixtures to make sure they're not broken.)
I agree with the factory approach -- creating fixtures on the fly, especially when the relationships are as complicated as they are here.